CA2229 (Implement serialization constructors) reported for delegate declaration

#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

SerializationRulesDiagnosticAnalyzer

#### Repro steps

```
using System;

class A
{
    [Serializable]
    public delegate void B();
}
```

#### Expected behavior
No CA2229

#### Actual behavior
`warning CA2229: Add a constructor to B with the following signature: 'protected B(SerializationInfo info, StreamingContext context)'.`

--- a/src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs
+++ b/src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs
@@ -155,6 +155,10 @@ namespace Microsoft.NetFramework.Analyzers
             public void AnalyzeSymbol(SymbolAnalysisContext context)
             {
                 var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+                if (namedTypeSymbol.TypeKind == TypeKind.Delegate || namedTypeSymbol.TypeKind == TypeKind.Interface)
+                {
+                    return;
+                }
 
                 // If the type is public and implements ISerializable
                 if (namedTypeSymbol.DeclaredAccessibility == Accessibility.Public && namedTypeSymbol.AllInterfaces.Contains(_iserializableTypeSymbol))

--- a/src/Microsoft.NetFramework.Analyzers/UnitTests/MarkISerializableTypesWithSerializableTests.cs
+++ b/src/Microsoft.NetFramework.Analyzers/UnitTests/MarkISerializableTypesWithSerializableTests.cs
@@ -222,13 +222,13 @@ Imports System
 Imports System.Runtime.Serialization
 
 Class A
-	<Serializable> _
-	Public Delegate Sub B()
+    <Serializable> _
+    Public Delegate Sub B()
 End Class
 
 Public Interface I
-	Inherits ISerializable
-	ReadOnly Property Name() As String
+    Inherits ISerializable
+    ReadOnly Property Name() As String
 End Interface");
         }
 

--- a/src/Microsoft.NetFramework.Analyzers/UnitTests/MarkISerializableTypesWithSerializableTests.cs
+++ b/src/Microsoft.NetFramework.Analyzers/UnitTests/MarkISerializableTypesWithSerializableTests.cs
@@ -199,6 +199,39 @@ namespace Microsoft.NetFramework.Analyzers.UnitTests
                 GetCA2237BasicResultAt(4, 30, "CA2237SerializableWithBaseAttr"));
         }
 
+        [Fact]
+        public void CA2237_CA2229_NoDiagnosticForInterfaceAndDelegate()
+        {
+            VerifyCSharp(@"
+using System;
+using System.Runtime.Serialization;
+
+class A
+{
+    [Serializable]
+    public delegate void B();
+}
+
+public interface I : ISerializable
+{
+    string Name { get; }
+}");
+
+            VerifyBasic(@"
+Imports System
+Imports System.Runtime.Serialization
+
+Class A
+	<Serializable> _
+	Public Delegate Sub B()
+End Class
+
+Public Interface I
+	Inherits ISerializable
+	ReadOnly Property Name() As String
+End Interface");
+        }
+
         internal static readonly string CA2237Message = MicrosoftNetFrameworkAnalyzersResources.MarkISerializableTypesWithSerializableMessage;
 
         private static DiagnosticResult GetCA2237CSharpResultAt(int line, int column, string objectName)

